packages/autocomplete-core/src/utils/createCancelablePromiseList.ts

```diff
@@ -0,0 +1,31 @@
+import { noop } from '@algolia/autocomplete-shared';
+
+import { CancelablePromise } from '.';
+
+export type CancelablePromiseList<TValue> = {
+  add(cancelablePromise: CancelablePromise<TValue>): void;
+  cancelAll(): void;
+  isEmpty(): boolean;
+};
+
+export function createCancelablePromiseList<
+  TValue
+>(): CancelablePromiseList<TValue> {
+  let list: Array<CancelablePromise<TValue>> = [];
+
+  return {
+    add(cancelablePromise) {
+      list.push(cancelablePromise);
+
+      cancelablePromise.catch(noop).finally(() => {
```

Comment:
does this line cause genuine errors to be swallowed? do we have a test for what happens if eg. algolia returns an error?

Reply:
Hard truth is, we don't handle errors for now and don't let users handle them as `onInput` is internal. We should likely expose some `onError` handler on sources and/or items, but this is a different topic.

But indeed, this `catch` would make errors not bubble up. We have [a test](https://github.com/algolia/autocomplete/blob/2dd34e0ac1eae19d87105668bd13155b543ca336/packages/autocomplete-core/src/__tests__/getSources.test.ts#L142-L167) that should have failed but it didn't because we have to rely on internals.

This `catch` was intended to avoid unhandled promise rejections during tests. Do you have ideas?

Reply:
could we catch outside of the function in the test instead of catching in the implementation? I see that this was already a problem before, but I wonder if this PR makes the situation more complicated?

Reply:
That's the thing, we can catch in the test, but having this consumer in l. 20 with only a `finally` handler makes Node complain. Having one consumer catch isn't enough.

Reply:
A solution would be not to rely on `finally`. The original implementation had an `onCancel` callback in the executor (relying on the same internals as `finally`), we could reintroduce it.

Reply:
[I changed the implementation of `createCancelablePromiseList` a bit](https://github.com/algolia/autocomplete/pull/864/commits/c14844cb0df7e4b6e048dcc6e239df543170c092) so that `add` returns the promise chain and we avoid creating lone, untracked promises. This fixes the problem and errors can bubble up (try typing a query in this sandbox and see the invariant violation throw in the console).

Unfortunately this isn't really [testable](https://github.com/algolia/autocomplete/blob/2dd34e0ac1eae19d87105668bd13155b543ca336/packages/autocomplete-core/src/__tests__/getSources.test.ts#L142-L168) at the moment because we don't have access to the promise chain from the outside (we don't return the chain from `onInput` calls). Our reliance on internals cause false negatives.

I suggest we tackle this problem (which was already there) in a dedicated error-handling feature.